adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1293 @@
+
+	var imageTitle = ""
+	imageTitle = getDecodeValue(content.MetaData.Title)
+	// dspImp, Imp, dspClick, Click tracking all can be found in MonitorUrl(imp ,click)
+	dspImpTrackings, dspClickTrackings := getDspImpClickTrackings(content)
+	var dspImpTrackings2StrImg strings.Builder
+	for i := 0; i < len(dspImpTrackings); i++ {
+		dspImpTrackings2StrImg.WriteString(`<img height="1" width="1" src='`)
+		dspImpTrackings2StrImg.WriteString(dspImpTrackings[i])
+		dspImpTrackings2StrImg.WriteString(`' >  `)
+	}
+
+	adm = "<style> html, body  " +
+		"{ margin: 0; padding: 0; width: 100%; height: 100%; vertical-align: middle; }  " +
+		"html  " +
+		"{ display: table; }  " +
+		"body { display: table-cell; vertical-align: middle; text-align: center; -webkit-text-size-adjust: none; }  " +
+		"</style> " +
+		`<span class="title-link advertiser_label">` + imageTitle + "</span> " +
+		"<a href='" + clickUrl + `' style="text-decoration:none" ` +
+		"onclick=sendGetReq()> " +
+		"<img src='" + imageInfoUrl + "' width='" + strconv.Itoa(int(adWidth)) + "' height='" + strconv.Itoa(int(adHeight)) + "'/> " +
+		"</a> " +
+		dspImpTrackings2StrImg.String() +
+		`<script type="text/javascript">` +
+		"var dspClickTrackings = [" + dspClickTrackings + "];" +
+		"function sendGetReq() {" +
+		"sendSomeGetReq(dspClickTrackings)" +
+		"}" +
+		"function sendOneGetReq(url) {" +
+		"var req = new XMLHttpRequest();" +
+		"req.open('GET', url, true);" +
+		"req.send(null);" +
+		"}" +
+		"function sendSomeGetReq(urls) {" +
+		"for (var i = 0; i < urls.length; i++) {" +
+		"sendOneGetReq(urls[i]);" +
+		"}" +
+		"}" +
+		"</script>"
+	return adm, adWidth, adHeight, nil
+}
+
+func getDspImpClickTrackings(content *content) (dspImpTrackings []string, dspClickTrackings string) {
+	for _, monitor := range content.Monitor {
+		if monitor.EventType == "imp" && len(monitor.Url) != 0 {
+			dspImpTrackings = monitor.Url
+		}
+		if monitor.EventType == "click" && len(monitor.Url) != 0 {
+			dspClickTrackings = getStrings(monitor.Url)
+		}
+	}
+	return dspImpTrackings, dspClickTrackings
+}
+
+func getStrings(eles []string) string {
+	if len(eles) == 0 {
+		return ""
+	}
+	var strs strings.Builder
+	for i := 0; i < len(eles); i++ {
+		strs.WriteString("\"" + eles[i] + "\"")
+		if i < len(eles)-1 {
+			strs.WriteString(",")
+		}
+	}
+	return strs.String()
+}
+
+// getDuration: millisecond -> format: 00:00:00.000
+func getDuration(duration int64) string {
+	// duration millisecond
+	if duration == 0 {
+		return "00:00:00.000"
+	}
+	var totalSec = int(duration) / 1000
+	var mmm = int(duration) % 1000
+	var hour = totalSec / 3600
+	var tmp = totalSec % 3600
+	var min = tmp / 60
+	var sec = tmp % 60
+	var result strings.Builder
+	result.WriteString(addZero(strconv.Itoa(hour), ":", false))
+	result.WriteString(addZero(strconv.Itoa(min), ":", false))
+	result.WriteString(addZero(strconv.Itoa(sec), ".", false))
+	result.WriteString(addZero(strconv.Itoa(mmm), "", true))
+	return result.String()
+}
```

Comment:
I don't have a problem with `getDuration(duration int64)`, it looks pretty nice and I don't mind leaving it as is. However, golang's `"time"` package seems to be able to do something similar. An alternative to `getDuration` solely with tools provided in the `"time"` package could probably be:
```
func getDurationWithGolangTools(dur int64) string {
    var d time.Duration = time.Duration(dur) * time.Millisecond
    t := time.Time{}.Add(d)

    return t.Format("15:04:05.000")
}
```
The following test case seems to match the return values of both functions:
```
func TestFunc(t *testing.T) {
	alternativeFunc := func(dur int64) string {
		var d time.Duration = time.Duration(dur) * time.Millisecond
		t := time.Time{}.Add(d)

		return t.Format("15:04:05.000")
	}

	testCases := []struct {
		desc     string
		duration int64
	}{
		{
			"01:03:20.001",
			3800001,
		},
		{
			"13:03:20.001",
			3600*13 + 3*60 + 20001,
		},
		{
			"twenty five hours",
			25 * 1000,
		},
		{
			"23:59:59.000",
			23*3600*1000 + 3599*1000,
		},
	}
	for _, tc := range testCases {
		// current duration function
		curr := getDuration(tc.duration)

		// duration function using golang libraries
		proposed := alternativeFunc(tc.duration)

		assert.Equal(t, curr, proposed)
	}
}
```
This is just a suggestion to maintain less code. Feel free to leave `getDuration(duration int64)` as is If you think it's more convenient.

Reply:
Awesome. Just in case I missed and edge test case, make sure the new function is 100% equivalent to yours.

Reply:
yeah, I have tested.